## COSC 1210
### print(), input(), value storage and chaos

Objectives
- use the chaos function to illustrate the importance of value storage

In [1]:

def main():
    a = 1/7
    b = a + a + a + a + a + a + a
    print(b)
main()

0.9999999999999998


##### Explain why the function above provides an answer that is different than the algebraic solution


In [ ]:
# because python doesn't store infinite values and round the floats when adding the numbers.

In [2]:
def main():
    a = 1/7
    b = a * 7 # has built-in rounding 
    c = a + a + a + a + a + a + a
    print("b is", b)
    print("c is", c)
main()

b is 1.0
c is 0.9999999999999998


##### Explain why b and c provide different answers



##### How large is the difference between c and b?




In [ ]:
# because python can just multiply it by 7 and get the answer and it doesn't need to round the values, while in c it needs to convert 1/7 into decimals and add it 7 times.

In [ ]:
# the difference is very small around 0.0000000000000000000000000001

### rounding floats
Mathematically, what will be the value of b?
When calculated by python, what is the value of b?

In [10]:
def main():
    print("This program illustrates a chaotic function")
    iterations = 10
    x1 = 0.15
    for i in range(iterations):
        x1 = 3.9 * x1 * (1 - x1)
    print("For", iterations, "iterations, x1 is:", x1)
main()



This program illustrates a chaotic function
For 10 iterations, x1 is: 0.9735732406265619


In [12]:
def main():
    print("This program illustrates a chaotic function")
    iterations = 100
    x1 = 0.15
    for i in range(iterations):
        x1 = 3.9 * x1 * (1 - x1)
    print("For", iterations, "iterations, x1 is:", x1)
main()


This program illustrates a chaotic function
For 100 iterations, x1 is: 0.5126527786207501


##### If you calculate x1 for 10 iterations vs 100 iterations, what answers do you get?

for 10 iterations: x1 is: 0.9735732406265619

for 100 iterations: x1 is: 0.5126527786207501

##### Why are these different?
the number changes each time through the calculation


In [8]:
def main():
    print("This program illustrates a chaotic function")
    iterations = 10
    x1 = 0.15
    x2 = 0.15
    for i in range(iterations):
        x1 = 3.9 * x1 * (1 - x1)
        x2 = 3.9 * (x2 - x2 * x2)
    print("For", iterations, "iternations, x1 is now", x1, "and x2 is now", x2)
main()

This program illustrates a chaotic function
For 10 iternations, x1 is now 0.9735732406265619 and x2 is now 0.9735732406265659


In [13]:
def main():
    print("This program illustrates a chaotic function")
    iterations = 100
    x1 = 0.15
    x2 = 0.15
    for i in range(iterations):
        x1 = 3.9 * x1 * (1 - x1)
        x2 = 3.9 * (x2 - x2 * x2)
    print("For", iterations, "iternations, x1 is now", x1, "and x2 is now", x2)
main()

This program illustrates a chaotic function
For 100 iternations, x1 is now 0.5126527786207501 and x2 is now 0.16411537941360743


In [15]:
def main():
    print("This program illustrates a chaotic function")
    iterations = 10000
    x1 = 0.15
    x2 = 0.15
    for i in range(iterations):
        x1 = 3.9 * x1 * (1 - x1)
        x2 = 3.9 * (x2 - x2 * x2)
    print("For", iterations, "iternations, x1 is now", x1, "and x2 is now", x2)
main()

This program illustrates a chaotic function
For 10000 iternations, x1 is now 0.4140723309159207 and x2 is now 0.6789638578779758


##### If you calculate x1 and x2 for 10 iterations vs 100 iterations, what answers do you get?
x1 10 iterations: 0.9735732406265619
x2 10 iterations: 0.9735732406265659

x1 100 iterations:0.5126527786207501
x2 100 iterations:0.16411537941360743

Why are these different?

In [ ]:
### Rewrite the chaos function so that the use can enter x
